### PR TITLE
chore: add a build tool that will produce docs for WITH options

### PR DESCRIPTION
**Pull request by neverchanje**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

The motivation of this tool is to help us maintain a mainifest file, a single source of truth `with_options.yaml`, for all options of connectors. The file will allow our doc team to easily grasp the key information of the newly-added options, which occupy 20% of total doc updates. It also provide ChatGPT a more structural information and enable it to generate better docs. In the future, we can host the information in a system table that allows users to query via SQL.

On the other hand, the QA team can leverage this yaml file to auto-generate test cases to see if options behave identically as we state in the document. For example, we can ensure that a required option, e.g, `stream` in NatsConfig, will return a friendly error when users don't specify it.

----

NEW UPDATE: 

Now the tool will also expand the connector option's default value. For example:

```yaml
  - name: properties.sync.call.timeout
    field_type: Duration
    required: false
    default: 'Duration :: from_secs (5)'
```

It also automatically infers whether a config is optional. For a Option, a HashMap, or a config with a default setter, it'll be considered optional. Otherwise required.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/connector/src/common.rs [hunk 1]
```diff
@@ -51,15 +52,17 @@ pub struct AwsPrivateLinkItem {
 }
 
 #[serde_as]
-#[derive(Debug, Clone, Serialize, Deserialize)]
+#[derive(Debug, Clone, Serialize, Deserialize, WithOptions)]
 pub struct KafkaCommon {
+    #[with_option(required)]
```
Comment (fuyufjh) [inline src/connector/src/common.rs:57]: Can we derive this information from `Option<>` or not? i.e. Optional arguments are not required, and vice versa.

File: src/connector/with_options.yaml [hunk 2]
```diff
@@ -0,0 +1,729 @@
...
+    required: false
+  - name: properties.enable.auto.commit
+    field_type: Option < bool >
+    comments: 'Automatically and periodically commit offsets in the background.  Note: setting this to false does not prevent the consumer from fetching previously committed start offsets.  To circumvent this behaviour set specific start offsets per partition in the call to assign().  default: true'
+    required: false
+KinesisProperties:
+  fields:
+  - name: scan.startup.mode
+    field_type: Option < String >
+    required: false
```
Comment (fuyufjh) [inline src/connector/with_options.yaml:354]: Lots of fields are not well-documents. Need to be improved later.

**Review by fuyufjh — COMMENTED** [review]
LGTM. Didn't carefully check the parameters for each connector. Maybe @tabVersion can help.

File: src/utils/with_options/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+serde = "1"
+serde_yaml = "0.9"
+syn = { version = "2", features = ["full"] }
+walkdir = "2"
+
+[lib]
+proc-macro = true
+
+[target.'cfg(not(madsim))'.dependencies]
+workspace-hack = { path = "../../workspace-hack" }
```
Comment (BugenZhao) [inline src/utils/with_options/Cargo.toml:23]:
```suggestion
```

Seems unnecessary for proc-macro crates? cc @xxchan

Comment (xxchan) [reply src/utils/with_options/Cargo.toml:23]: Yes, we should not do it.
https://github.com/risingwavelabs/risingwave/pull/12961

File: src/connector/with_options.yaml
Comment (xxchan) [inline src/connector/with_options.yaml:1]: 1. We'd better have a comment here mentioning this file is auto-generated (and how is it generated)

2. Should we check in this file? (I'm ok with both, but slightly prefer not)

   If we do, we'd better check in CI that it's up to date.

Comment (xxchan) [reply src/connector/with_options.yaml:1]: Found a prior art

https://github.com/risingwavelabs/risingwave/blob/624094cc28d16932425f7c41733285bffd51a15a/src/common/src/config.rs#L1483-L1500

File: src/connector/build.rs [hunk 4]
```diff
@@ -13,10 +13,11 @@
 // limitations under the License.
 
 fn main() {
-    let proto_dir = "./src/test_data/proto_recursive";
-
-    println!("cargo:rerun-if-changed={}", proto_dir);
+    // This function extracts WITH options from the Rust code and maintains
+    // them in a single source of truth `with_options.yaml`.
+    with_options_build::update_with_options_yaml();
```
Comment (xxchan) [inline src/connector/build.rs:18]: Why to put this in `build.rs`? Not strongly opposed to this, but feel a little strange. A bin might be more natural (like `generate-example-config`).

Comment (xxchan) [reply src/connector/build.rs:18]: Oh, we can do a `expect-test`, which can be used to both generate the file, and ensure it's up-to-date.

Comment (BugenZhao) [reply src/connector/build.rs:18]: Now the build script will rerun if there's any change in this crate. Have you checked the impact on the developer experience?

File: src/connector/src/common.rs (hunk 1 above)
Comment (BugenZhao) [reply src/connector/src/common.rs:57]: Could you provide an example for this?

**Comment by tabVersion** [comment]
> LGTM. Didn't carefully check the parameters for each connector. Maybe @tabVersion can help.

The essential props for mq connectors are basically correct, need more time verifying others.
shall we make the yml reference some config blocks? For example, if we change a field in `KafkaCommon`, we need to do the same change twice for both kafka source and kafka sink in the yml file.

**Review by xxhZs — APPROVED** [review]

File: src/connector/src/sink/doris.rs [hunk 5]
```diff
@@ -38,8 +39,9 @@ use crate::sink::{
 
 pub const DORIS_SINK: &str = "doris";
 
-#[derive(Deserialize, Serialize, Debug, Clone)]
+#[derive(Deserialize, Serialize, Debug, Clone, WithOptions)]
 pub struct DorisCommon {
+    #[with_option(required)]
     #[serde(rename = "doris.url")]
     pub url: String,
     #[serde(rename = "doris.user")]
```
Comment (xxhZs) [inline src/connector/src/sink/doris.rs:47]: They are all required

**Comment by tabVersion** [comment]
Hi @neverchanje, just for clarification, is the pr aiming to eliminate all hashmaps as the backend of `with` clause?
Had an offline discussion with @StrikeW and he told me there are a bunch of dbz params to be passed to embedded dbz. These params are not documented because they are available in dbz's doc

File: src/connector/build.rs (hunk 4 above)
Comment (BugenZhao) [reply src/connector/build.rs]: LGTM!

**Review by tabVersion — APPROVED** [review]
No more question. Thanks for your effort 👍

**Review by xxchan — APPROVED** [review]

File: src/connector/with_options.yaml [hunk 6]
```diff
@@ -0,0 +1,835 @@
...
+    field_type: String
+    required: true
+  - name: s3.secret.key
+    field_type: String
+    required: true
+  - name: primary_key
+    field_type: Option < Vec < String > >
+    required: false
+    default: Default::default
+KafkaConfig:
```
Comment (xxchan) [inline src/connector/with_options.yaml:107]: KafkaConfig is source, while KafkaProperties is sink. This is confusing. Maybe we can add a `type` field, and/or use two files for source and sink

**Pull request merged**